File: go/pass6.go
package main

import (
	"strings"
)

// Check if a node is a leaf
func (i *Ifcomp) leaf(n TreeIndex) bool {
	return i.TreeState.Node[n].BranchStart == NullNode
}

// Get absolute line number from node
func (i *Ifcomp) trueLineOf(N TreeIndex) LineCount {
	if i.TreeState.Node[N].Linen < 0 {
		return -i.TreeState.Node[N].Linen
	}
	return i.TreeState.Node[N].Linen
}

// Print formatted header
func (i *Ifcomp) printHeader(s string) {
	padding := 52
	out := s + " " + strings.Repeat("=", padding-len(s))
	i.printf("*** %s ***\n", out)
}

// Print formatted header1
func (i *Ifcomp) printHeader1(s string) {
	padding := 52
	out := s + " " + strings.Repeat("-", padding-len(s))
	i.printf("*** %s ***\n", out)
}

// Print formatted trailer
func (i *Ifcomp) printTrailer() {
	out := strings.Repeat("=", 53)
	i.printf("*** %s ***\n\n", out)
}

// Print lines in a node
func (i *Ifcomp) printNode(noden TreeIndex) {
	i.printNode1(noden, false, 0)
}

// Print lines in a node (with options)
func (i *Ifcomp) printNode1(noden TreeIndex, always bool, startingLine int) {
	i.eachLineInNode(noden, always, startingLine, func(whichFile FileIndex, text string, lineno int) {
		prefix := " "
		if whichFile == Second {
			prefix = "+"
		}
		i.printf("%s%6d|%s\n", prefix, lineno, text)
	})
}

// Iterate through all lines in a node
func (i *Ifcomp) eachLineInNode(noden TreeIndex, always bool, startingLine int, fn func(FileIndex, string, int)) {
	var start, finish TreeIndex
	if !i.leaf(noden) {
		start = i.TreeState.Node[noden].BranchStart
		finish = noden
	} else {
		start = noden
		finish = i.TreeState.Node[noden].Next
	}

	for current := start; current != finish; current = i.TreeState.Node[current].Next {
		sline := i.TreeState.Node[current].Linen
		fileno := getWhichFile(sline)
		sline = getAbsLine(sline)
		filenoIdx := toArrayIndex(fileno)

		// cost is the number of nodes. Can be negative.
		cost := int(i.TreeState.Node[current].Cost)
		if always {
			if cost < 0 {
				cost = -cost
			}
		}
		last := int(sline) + cost

		// He may have passed a place to start later than the beginning of a node.
		maxStart := int(sline)
		if startingLine > maxStart {
			maxStart = startingLine
		}
		for s := maxStart; s < last; s++ {
			fileno2 := FileIndex(fileno)
			text := i.LineMatchingState.StringTable[i.FileState.FileLine[filenoIdx][s].FileLineText].Text
			lineno := int(i.FileState.FileLine[filenoIdx][s].Linen)
			fn(fileno2, text, lineno)
		}
	}
}

// Count cosmetic and non-cosmetic lines in a node
func (i *Ifcomp) countNode(noden TreeIndex, p *LineKinds) {
	i.eachLineInNode(noden, false, 0, func(whichFile FileIndex, text string, lineno int) {
		// For now, no cosmetic lines (cosmetic_line always returns false in C++)
		p.NonCosmetic++
	})
}

// Remove a node from its linked list
func (i *Ifcomp) detachNode(noden TreeIndex) {
	prev := i.TreeState.Node[noden].Prev
	next := i.TreeState.Node[noden].Next
	i.TreeState.Node[prev].Next = next
	i.TreeState.Node[next].Prev = prev
}

// Find node in tree containing the specified line number
func (i *Ifcomp) findNode(T TreeBounds, linen TreeIndex) TreeIndex {
	absLinen := linen
	if linen < 0 {
		absLinen = -linen
	}
	N := T.Start
	for N != T.End {
		if i.trueLineOf(N) == LineCount(absLinen) {
			if i.DebugDumpTreesFull {
				i.printf("In tree %d:%d, find line %d at %d\n", T.Start, T.End, linen, N)
			}
			return N
		}
		N = i.TreeState.Node[N].Next
	}
	// Node not found
	if i.DebugDumpTreesFull {
		N = T.Start
		i.printf("[")
		for N != T.End {
			i.printf("%d ", N)
			N = i.TreeState.Node[N].Next
		}
		i.printf("] ln=%d\n", linen)
		i.printf("*** Warning: find_node could not find line %d in tree %d:%d\n", linen, T.Start, T.End)
	}
	return NullNode
}

// Check if an unmatched node in file1 can be replaced
func (i *Ifcomp) pass6Replaceable(noden TreeIndex) TreeIndex {
	firstIdx := toArrayIndex(First)
	secondIdx := toArrayIndex(Second)

	prev := i.TreeState.Node[noden].Prev
	prevOtherFile := i.findNode(i.TreeState.Trees[secondIdx], TreeIndex(i.FileState.FileLine[firstIdx][i.trueLineOf(prev)].Ptr0))

	if prevOtherFile == NullNode {
		return NullNode
	}

	nodenOtherFile := i.TreeState.Node[prevOtherFile].Next
	if i.TreeState.Node[nodenOtherFile].Cost >= 0 {
		if i.DebugDumpTreesFull {
			i.printf("replaceable fails: noden_other_file(%d) has neg cost.\n", nodenOtherFile)
		}
		return NullNode
	}
	return nodenOtherFile
}

// Process and output a deletion operation
func (i *Ifcomp) deleteLines(noden TreeIndex) {
	i.Stats.NChangeBlocks++
	i.afterHeader(i.TreeState.Node[noden].Prev)
	i.TreeState.Node[noden].Cost = -i.TreeState.Node[noden].Cost
	i.printHeader1("DELETE LINE(s)")
	i.printNode(noden)
	i.printTrailer()
	i.countNode(noden, &i.Stats.DeleteStats)
	i.detachNode(noden)
}

// Find first unique line in a node
func (i *Ifcomp) uniqueFind(noden TreeIndex) TreeIndex {
	endLine := i.TreeState.Node[noden].Linen
	filen := getWhichFile(endLine)
	endLine = getAbsLine(endLine)
	filenIdx := toArrayIndex(filen)

	cost := int(i.TreeState.Node[noden].Cost)
	for startLine := int(endLine) + cost - 1; startLine >= int(endLine); startLine-- {
		if i.FileState.FileLine[filenIdx][startLine].PtrType == UNIQUE_TYPE {
			return TreeIndex(startLine)
		}
	}
	return NullNode
}

// Print context lines before a change
func (i *Ifcomp) afterLines(noden TreeIndex) {
	i.printHeader("AFTER LINE(s)")
	firstIdx := toArrayIndex(First)
	start := noden
	last := i.TreeState.Node[start].Next
	var linen TreeIndex = 0

	for start != i.TreeState.Trees[firstIdx].Start {
		if i.leaf(start) {
			linen = i.uniqueFind(start)
			if linen != NullNode {
				break
			}
			linen = 0
			last = start
			start = i.TreeState.Node[start].Prev
		} else {
			if last == i.TreeState.Node[start].BranchStart {
				start = i.TreeState.Node[start].Prev
				last = start
			} else {
				last = start
				start = i.TreeState.Node[start].BranchEnd
			}
		}
	}

	i.printNode1(start, false, int(linen))
	last = start
	start = i.TreeState.Node[start].Next

	for start != i.TreeState.Node[noden].Next {
		if i.leaf(start) {
			i.printNode(start)
			last = start
			start = i.TreeState.Node[start].Next
		} else {
			if last == i.TreeState.Node[start].BranchEnd {
				last = start
				start = i.TreeState.Node[start].Next
			} else {
				last = start
				start = i.TreeState.Node[start].BranchStart
			}
		}
	}
}

// Print context header
func (i *Ifcomp) afterHeader(noden TreeIndex) {
	firstIdx := toArrayIndex(First)
	if noden == i.TreeState.Trees[firstIdx].Start {
		i.printHeader("AFTER TOP")
	} else {
		i.afterLines(noden)
	}
}

// Process and output a replacement operation
func (i *Ifcomp) pass6ReplaceLines(node1, node2 TreeIndex) {
	i.Stats.NChangeBlocks++
	// Make the costs positive, indicating that the nodes now
	// correspond to something in the other file.
	i.TreeState.Node[node1].Cost = -i.TreeState.Node[node1].Cost
	i.TreeState.Node[node2].Cost = -i.TreeState.Node[node2].Cost
	i.countNode(node1, &i.Stats.Replace1Stats)
	i.countNode(node2, &i.Stats.Replace2Stats)
	i.afterHeader(i.TreeState.Node[node1].Prev)
	i.printHeader1("REPLACE LINE(s)")
	i.printNode(node1)
	i.printHeader1("WITH LINE(s)")
	i.printNode(node2)
	i.printTrailer()
	i.detachNode(node1)
	i.detachNode(node2)
}

// Process and output an insertion operation
func (i *Ifcomp) pass6InsertLines(noden TreeIndex) {
	firstIdx := toArrayIndex(First)
	secondIdx := toArrayIndex(Second)
	i.Stats.NChangeBlocks++
	i.TreeState.Node[noden].Cost = -i.TreeState.Node[noden].Cost
	i.countNode(noden, &i.Stats.InsertStats)
	prev := i.TreeState.Node[noden].Prev
	if prev == i.TreeState.Trees[secondIdx].Start {
		i.detachNode(noden)
		i.TreeState.Node[i.TreeState.Trees[firstIdx].Start].BranchStart = noden
		i.TreeState.Node[i.TreeState.Trees[firstIdx].Start].BranchEnd = noden
		i.TreeState.Node[noden].Prev = i.TreeState.Trees[firstIdx].Start
		i.TreeState.Node[noden].Next = i.TreeState.Trees[firstIdx].Start
		i.printHeader("AFTER TOP")
		i.printHeader1("INSERT LINE(s)")
		i.printNode(i.TreeState.Trees[firstIdx].Start)
	} else {
		j := i.findNode(i.TreeState.Trees[firstIdx],
			TreeIndex(i.FileState.FileLine[secondIdx][i.trueLineOf(prev)].Ptr0))
		if j == NullNode {
			i.detachNode(noden)
			i.freeNode(noden)
			i.Stats.NChangeBlocks--
			return
		}
		i.afterLines(j)
		i.printHeader1("INSERT LINE(s)")
		i.printNode(noden)
		i.combineNodes(j, noden)
	}
	i.printTrailer()
}

// Dump trees for debugging
func (i *Ifcomp) dumpTrees(pass int) {
	if !i.DebugDumpTrees {
		return
	}
	if pass == 99 {
		i.println("dump trees")
	} else {
		i.printf("dump_trees after pass%d\n", pass)
	}
	i.dumpTree(i.TreeState.Trees[0].Start)
	i.dumpTree(i.TreeState.Trees[1].Start)
}

// Dump tree structure for debugging
func (i *Ifcomp) dumpTree(treeStart TreeIndex) {
	i.printf("Tree %d:\n", treeStart)
	branch := false
	T := treeStart
	for T != NullNode {
		T2 := T
		if i.leaf(T) {
			pad := 0
			if branch {
				pad = 1
			}
			i.formatNode(T, pad)
			T = i.TreeState.Node[T].Next
			if i.DebugDumpTreesFull {
				i.printNode1(T2, true, 0)
			}
		} else {
			if branch {
				branch = false
				T = i.TreeState.Node[T].Next
			} else {
				i.formatNode(T, 0)
				T = i.TreeState.Node[T].BranchStart
				branch = true
			}
		}
	}
}

// Format and print node information
func (i *Ifcomp) formatNode(noden TreeIndex, pad int) {
	i.printf("%s[%d<-N%d->%d, cost=%2d linen=%2d",
		strings.Repeat(" ", pad*7), i.TreeState.Node[noden].Prev, noden,
		i.TreeState.Node[noden].Next, i.TreeState.Node[noden].Cost, i.TreeState.Node[noden].Linen)

	L := i.TreeState.Node[noden].Linen
	fileno := getWhichFile(L)
	L = getAbsLine(L)
	filenoIdx := toArrayIndex(fileno)

	// Check bounds before accessing FileLine - end nodes may have Linen beyond file bounds
	if int(L) < len(i.FileState.FileLine[filenoIdx]) {
		i.printf("(%d)", i.FileState.FileLine[filenoIdx][L].Ptr0)
	} else {
		i.printf("(end)")
	}

	if i.TreeState.Node[noden].BranchStart != NullNode || i.TreeState.Node[noden].BranchEnd != NullNode {
		i.printf(" bs=%2d be=%2d", i.TreeState.Node[noden].BranchStart, i.TreeState.Node[noden].BranchEnd)
	}
	i.printf("]\n")
}

// Make a new tree node
func (i *Ifcomp) makeNode(p NodeDecl) TreeIndex {
	i.TreeState.Node = append(i.TreeState.Node, p)
	idx := TreeIndex(len(i.TreeState.Node) - 1)
	if i.DebugDumpTreesFull {
		i.printf("just made ")
		i.formatNode(idx, 0)
	}
	return idx
}

// Pass 6: Replace/Delete/Insert operations
func (i *Ifcomp) pass6() {
	firstIdx := toArrayIndex(First)
	for j := i.TreeState.Node[i.TreeState.Trees[firstIdx].Start].Next; j != i.TreeState.Trees[firstIdx].End; {
		k := j
		j = i.TreeState.Node[j].Next
		if i.TreeState.Node[k].Cost < 0 {
			// Try to replace first
			l := i.pass6Replaceable(k)
			if l != NullNode {
				i.pass6ReplaceLines(k, l)
			} else {
				i.deleteLines(k)
			}
		}
	}

	secondIdx := toArrayIndex(Second)
	for j := i.TreeState.Node[i.TreeState.Trees[secondIdx].Start].Next; j != i.TreeState.Trees[secondIdx].End; {
		k := j
		j = i.TreeState.Node[j].Next
		if i.TreeState.Node[k].Cost < 0 {
			i.pass6InsertLines(k)
		}
	}
}

File: rust/src/pass6.rs
use crate::types::*;
use super::Ifcomp;

impl Ifcomp {
    // Remove a node from its linked list
    pub fn detach_node(&mut self, noden: TreeIndex) {
        let prev = self.tree_state.node[noden].prev;
        let next = self.tree_state.node[noden].next;
        self.tree_state.node[prev].next = next;
        self.tree_state.node[next].prev = prev;
    }

    // Find node in tree containing the specified line number
    // Note: linen can be negative for file2 line numbers, so we use i16
    pub fn find_node(&self, t: TreeBounds, linen: i16) -> TreeIndex {
        let abs_linen = if linen < 0 { -linen } else { linen };
        let mut n = t.start;
        while n != t.end {
            if self.true_line_of(n) == abs_linen as LineCount {
                if self.debug_dump_trees_full {
                    println!("In tree {}:{}, find line {} at {}", t.start, t.end, linen, n);
                }
                return n;
            }
            n = self.tree_state.node[n].next;
        }
        // Node not found
        if self.debug_dump_trees_full {
            let mut n2 = t.start;
            print!("[");
            while n2 != t.end {
                print!("{} ", n2);
                n2 = self.tree_state.node[n2].next;
            }
            println!("] ln={}", linen);
            println!("*** Warning: find_node could not find line {} in tree {}:{}", linen, t.start, t.end);
        }
        NULL_NODE
    }

    // Check if an unmatched node in file1 can be replaced
    pub fn pass6_replaceable(&self, noden: TreeIndex) -> TreeIndex {
        let first_idx = FileIndex::First.to_array_index();
        let second_idx = FileIndex::Second.to_array_index();

        let prev = self.tree_state.node[noden].prev;
        let prev_other_file = self.find_node(
            self.tree_state.trees[second_idx],
            self.file_state.file_line[first_idx][self.true_line_of(prev) as usize].ptr0 as i16
        );

        if prev_other_file == NULL_NODE {
            return NULL_NODE;
        }

        let noden_other_file = self.tree_state.node[prev_other_file].next;
        if self.tree_state.node[noden_other_file].cost >= 0 {
            if self.debug_dump_trees_full {
                println!("replaceable fails: noden_other_file({}) has neg cost.", noden_other_file);
            }
            return NULL_NODE;
        }
        noden_other_file
    }

    // Find first unique line in a node
    pub fn unique_find(&self, noden: TreeIndex) -> TreeIndex {
        let mut end_line = self.tree_state.node[noden].linen;
        let filen = get_which_file(end_line);
        end_line = get_abs_line(end_line);
        let filen_idx = filen.to_array_index();

        let cost = self.tree_state.node[noden].cost;
        let mut start_line = (end_line + cost - 1) as i32;
        let end_line_i32 = end_line as i32;
        while start_line >= end_line_i32 {
            if self.file_state.file_line[filen_idx][start_line as usize].ptr_type == LineType::UniqueType {
                return start_line as TreeIndex;
            }
            start_line -= 1;
        }
        NULL_NODE
    }

    // Print context lines before a change
    pub fn after_lines(&mut self, noden: TreeIndex) {
        self.print_header("AFTER LINE(s)");
        let first_idx = FileIndex::First.to_array_index();
        let mut start = noden;
        let mut last = self.tree_state.node[start].next;
        let mut linen: TreeIndex = 0;

        while start != self.tree_state.trees[first_idx].start {
            if self.leaf(start) {
                linen = self.unique_find(start);
                if linen != NULL_NODE {
                    break;
                }
                linen = 0;
                last = start;
                start = self.tree_state.node[start].prev;
            } else {
                if last == self.tree_state.node[start].branch_start {
                    start = self.tree_state.node[start].prev;
                    last = start;
                } else {
                    last = start;
                    start = self.tree_state.node[start].branch_end;
                }
            }
        }

        self.print_node1(start, false, linen as usize);
        last = start;
        start = self.tree_state.node[start].next;

        while start != self.tree_state.node[noden].next {
            if self.leaf(start) {
                self.print_node(start);
                last = start;
                start = self.tree_state.node[start].next;
            } else {
                if last == self.tree_state.node[start].branch_end {
                    last = start;
                    start = self.tree_state.node[start].next;
                } else {
                    last = start;
                    start = self.tree_state.node[start].branch_start;
                }
            }
        }
    }

    // Print context header
    pub fn after_header(&mut self, noden: TreeIndex) {
        let first_idx = FileIndex::First.to_array_index();
        if noden == self.tree_state.trees[first_idx].start {
            self.print_header("AFTER TOP");
        } else {
            self.after_lines(noden);
        }
    }

    // Print formatted header
    pub fn print_header(&self, s: &str) {
        let padding: usize = 52;
        let repeat_count = padding.saturating_sub(s.len());
        let out = format!("{} {}", s, "=".repeat(repeat_count));
        println!("*** {} ***", out);
    }

    // Print formatted header1
    pub fn print_header1(&self, s: &str) {
        let padding: usize = 52;
        let repeat_count = padding.saturating_sub(s.len());
        let out = format!("{} {}", s, "-".repeat(repeat_count));
        println!("*** {} ***", out);
    }

    // Print formatted trailer
    pub fn print_trailer(&self) {
        let out = "=".repeat(53);
        println!("*** {} ***\n", out);
    }

    // Print lines in a node
    pub fn print_node(&mut self, noden: TreeIndex) {
        self.print_node1(noden, false, 0);
    }

    // Print lines in a node (with options)
    pub fn print_node1(&mut self, noden: TreeIndex, always: bool, starting_line: usize) {
        self.each_line_in_node(noden, always, starting_line, |which_file: FileIndex, text: &str, lineno: usize| {
            let prefix = if which_file == FileIndex::Second { "+" } else { " " };
            println!("{}{:6}|{}", prefix, lineno, text);
        });
    }

    // Process and output a deletion operation
    pub fn delete_lines(&mut self, noden: TreeIndex) {
        self.stats.n_change_blocks += 1;
        self.after_header(self.tree_state.node[noden].prev);
        self.tree_state.node[noden].cost = -self.tree_state.node[noden].cost;
        self.print_header1("DELETE LINE(s)");
        self.print_node(noden);
        self.print_trailer();
        let mut delete_kinds = LineKinds::default();
        self.count_node(noden, &mut delete_kinds);
        self.stats.delete_stats.cosmetic += delete_kinds.cosmetic;
        self.stats.delete_stats.non_cosmetic += delete_kinds.non_cosmetic;
        self.detach_node(noden);
    }

    // Process and output a replacement operation
    pub fn pass6_replace_lines(&mut self, node1: TreeIndex, node2: TreeIndex) {
        self.stats.n_change_blocks += 1;
        // Make the costs positive, indicating that the nodes now
        // correspond to something in the other file.
        self.tree_state.node[node1].cost = -self.tree_state.node[node1].cost;
        self.tree_state.node[node2].cost = -self.tree_state.node[node2].cost;
        let mut replace1_kinds = LineKinds::default();
        let mut replace2_kinds = LineKinds::default();
        self.count_node(node1, &mut replace1_kinds);
        self.count_node(node2, &mut replace2_kinds);
        self.stats.replace1_stats.cosmetic += replace1_kinds.cosmetic;
        self.stats.replace1_stats.non_cosmetic += replace1_kinds.non_cosmetic;
        self.stats.replace2_stats.cosmetic += replace2_kinds.cosmetic;
        self.stats.replace2_stats.non_cosmetic += replace2_kinds.non_cosmetic;
        self.after_header(self.tree_state.node[node1].prev);
        self.print_header1("REPLACE LINE(s)");
        self.print_node(node1);
        self.print_header1("WITH LINE(s)");
        self.print_node(node2);
        self.print_trailer();
        self.detach_node(node1);
        self.detach_node(node2);
    }

    // Combine two adjacent nodes into a branch structure
    pub fn combine_nodes(&mut self, node1: TreeIndex, node2: TreeIndex) {
        let mut n = NodeDecl {
            cost: self.tree_state.node[node1].cost + self.tree_state.node[node2].cost,
            linen: self.tree_state.node[node1].linen,
            prev: NULL_NODE,
            next: NULL_NODE,
            branch_start: NULL_NODE,
            branch_end: NULL_NODE,
        };

        // First remove node2 from file2
        self.detach_node(node2);
        n.prev = self.tree_state.node[node1].prev;
        n.next = self.tree_state.node[node1].next;

        // Now remove node1 from file1
        self.detach_node(node1);

        let branch_link1: TreeIndex;
        let branch_link2: TreeIndex;

        if !self.leaf(node1) {
            // Just want the branch
            n.branch_start = self.tree_state.node[node1].branch_start;
            branch_link1 = self.tree_state.node[node1].branch_end;
            // The sequence in node1 is absorbed in N and hence isn't needed
            self.free_node(node1);
            // node1 is now replaced by n.branch_start conceptually
        } else {
            n.branch_start = node1;
            branch_link1 = node1;
        }

        if !self.leaf(node2) {
            branch_link2 = self.tree_state.node[node2].branch_start;
            n.branch_end = self.tree_state.node[node2].branch_end;
            // The sequence in node2 is absorbed in N and hence isn't needed
            self.free_node(node2);
        } else {
            branch_link2 = node2;
            n.branch_end = node2;
        }

        let new_node = self.make_node(n);
        // Insert new_node after N.prev and before N.next; i.e., it replaces node1
        self.tree_state.node[n.prev].next = new_node;
        self.tree_state.node[n.next].prev = new_node;
        self.tree_state.node[n.branch_start].prev = new_node;
        self.tree_state.node[n.branch_end].next = new_node;
        self.tree_state.node[branch_link1].next = branch_link2;
        self.tree_state.node[branch_link2].prev = branch_link1;
    }

    // Process and output an insertion operation
    pub fn pass6_insert_lines(&mut self, noden: TreeIndex) {
        let first_idx = FileIndex::First.to_array_index();
        let second_idx = FileIndex::Second.to_array_index();
        self.stats.n_change_blocks += 1;
        self.tree_state.node[noden].cost = -self.tree_state.node[noden].cost;
        let mut insert_kinds = LineKinds::default();
        self.count_node(noden, &mut insert_kinds);
        self.stats.insert_stats.cosmetic += insert_kinds.cosmetic;
        self.stats.insert_stats.non_cosmetic += insert_kinds.non_cosmetic;
        let prev = self.tree_state.node[noden].prev;
        if prev == self.tree_state.trees[second_idx].start {
            self.detach_node(noden);
            self.tree_state.node[self.tree_state.trees[first_idx].start].branch_start = noden;
            self.tree_state.node[self.tree_state.trees[first_idx].start].branch_end = noden;
            self.tree_state.node[noden].prev = self.tree_state.trees[first_idx].start;
            self.tree_state.node[noden].next = self.tree_state.trees[first_idx].start;
            self.print_header("AFTER TOP");
            self.print_header1("INSERT LINE(s)");
            self.print_node(self.tree_state.trees[first_idx].start);
        } else {
            let j = self.find_node(
                self.tree_state.trees[first_idx],
                self.file_state.file_line[second_idx][self.true_line_of(prev) as usize].ptr0 as i16
            );
            if j == NULL_NODE {
                self.detach_node(noden);
                self.free_node(noden);
                self.stats.n_change_blocks -= 1;
                return;
            }
            self.after_lines(j);
            self.print_header1("INSERT LINE(s)");
            self.print_node(noden);
            self.combine_nodes(j, noden);
        }
        self.print_trailer();
    }

    // Pass 6: Replace/Delete/Insert operations
    //
    // Purpose: Process unmatched segments by either replacing them with corresponding
    // segments from the other file, deleting them, or inserting them. This pass converts
    // negative-cost nodes (unmatched segments) into positive-cost nodes (matched segments)
    // or removes them from the tree structure.
    pub fn pass6(&mut self) {
        let first_idx = FileIndex::First.to_array_index();
        let mut j = self.tree_state.node[self.tree_state.trees[first_idx].start].next;
        while j != self.tree_state.trees[first_idx].end {
            let k = j;
            j = self.tree_state.node[j].next;
            if self.tree_state.node[k].cost < 0 {
                // Try to replace first
                let l = self.pass6_replaceable(k);
                if l != NULL_NODE {
                    self.pass6_replace_lines(k, l);
                } else {
                    self.delete_lines(k);
                }
            }
        }

        let second_idx = FileIndex::Second.to_array_index();
        let mut j = self.tree_state.node[self.tree_state.trees[second_idx].start].next;
        while j != self.tree_state.trees[second_idx].end {
            let k = j;
            j = self.tree_state.node[j].next;
            if self.tree_state.node[k].cost < 0 {
                self.pass6_insert_lines(k);
            }
        }
    }
}
